kaldb/src/main/java/com/slack/kaldb/zipkinApi/ZipkinService.java

```diff
@@ -57,8 +80,172 @@ public HttpResponse getTraces(
+          }
+        } else {
+          messageTags.put(k, String.valueOf(value));
+        }
+      }
+      // these are some mandatory fields without which the grafana zipkin plugin fails to display
+      // the span
+      if (messageTraceId == null) {
+        messageTraceId = message.id;
+      }
+      if (timestamp == null) {
+        LOG.warn("Document id={} missing @timestamp", message.id);
+        continue;
+      }
+
+      final long messageConvertedTimestamp = convertToMicroSeconds(Instant.parse(timestamp));
+
+      final Trace.ZipkinSpan span =
+          makeSpan(
+              messageTraceId,
+              Optional.ofNullable(parentId),
+              message.id,
+              Optional.ofNullable(name),
+              Optional.ofNullable(serviceName),
+              messageConvertedTimestamp,
+              duration,
+              messageTags);
+      String spanJson = printer.print(span);
+      traces.add(spanJson);
+    }
+    StringJoiner outputJsonArray = new StringJoiner(",", "[", "]");
+    traces.forEach(outputJsonArray::add);
+    return outputJsonArray.toString();
+  }
+
+  @VisibleForTesting
+  // Epoch microseconds of the start of this span, possibly absent if this an incomplete span
+  public static long convertToMicroSeconds(Instant instant) {
+    return ChronoUnit.MICROS.between(Instant.EPOCH, instant);
+  }
+
+  // We sadly need to create ZipkinSpan and can't reuse the Span proto
+  // The reason being, Span has id/traceId/parentId as bytes (which is correct and is how the proto
+  // definition is defined - there shouldn't be a string without encoding)
+  // However if we ship back the bytes ( encoded data ) back to grafana it has no idea that the data
+  // needs to decoded. Hence we decode it back and ship a ZipkinSpan
+  public static Trace.ZipkinSpan makeSpan(
+      String traceId,
+      Optional<String> parentId,
+      String id,
+      Optional<String> name,
+      Optional<String> serviceName,
+      long timestamp,
+      long duration,
+      Map<String, String> tags) {
+    Trace.ZipkinSpan.Builder spanBuilder = Trace.ZipkinSpan.newBuilder();
+
+    spanBuilder.setTraceId(ByteString.copyFrom(traceId.getBytes()).toStringUtf8());
+    spanBuilder.setId(ByteString.copyFrom(id.getBytes()).toStringUtf8());
+    spanBuilder.setTimestamp(timestamp);
+    spanBuilder.setDuration(duration);
+
+    parentId.ifPresent(
+        s -> spanBuilder.setParentId(ByteString.copyFrom(s.getBytes()).toStringUtf8()));
+    name.ifPresent(spanBuilder::setName);
+    serviceName.ifPresent(
+        s -> spanBuilder.setRemoteEndpoint(Trace.Endpoint.newBuilder().setServiceName(s)));
+    spanBuilder.putAllTags(tags);
+    return spanBuilder.build();
+  }
+
+  // intentionally returning LogWireMessage instead of LogMessage
+  // If we return LogMessage the caller then needs to call getSource which is a deep copy of the
+  // object
+  public static List<LogWireMessage> searchResultToLogWireMessage(
```

Comment:
SearchResult is a proto object. The result is coming from a gRPC response so I think this is the only way to convert it back to a LogWireMessage so that we can convert it to a ZipkinSpan

See https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/elasticsearchApi/ElasticsearchApiService.java#L118 which does the same thing currently

I don't know of a way to convert a proto to another proto without converting it. if you have a example code we can give it s shot